Support specifying namespace to generate classes in

LukeTowers opened the pull request:
This both fixes an infinite loop caused when the application namespace is an empty string (as it is in a domain-driven codebase like Winter CMS as there is no "root" namespace for the application) while also adding support for the --in option originally proposed in https://github.com/laravel/framework/pull/48571 by @innocenzi. 

This PR takes it a step further to resolve the correct path for real namespaces so that namespaces outside of root namespace (which may or may not actually exist) can also be written to.

LukeTowers committed:
Support specifying namespace to generate classes in

This both fixes an infinite loop caused when the application namespace is an empty string (as it is in a domain-driven codebase like Winter CMS as there is no "root" namespace for the application) while also adding support for the --in option originally proposed in https://github.com/laravel/framework/pull/48571 by @innocenzi

LukeTowers committed:
Add method to resolve class path using Composer

LukeTowers committed:
Style fix

LukeTowers committed:
Fix typo

shaedrich commented on src/Illuminate/Console/GeneratorCommand.php:
```diff
@@ -130,6 +137,13 @@ public function __construct(Filesystem $files)
+            null,
+            InputOption::VALUE_REQUIRED,
+            "Specify a namespace to generate the {$this->type} class in"
+        ));
```
Yeah, that's what `getOptions()` is for

LukeTowers commented on src/Illuminate/Console/GeneratorCommand.php:
```diff
@@ -433,7 +508,20 @@ protected function getNameInput()
+        }
+
+        return $this->rootNamespace = trim($in, '\\').'\\';
     }
```
That doesn't accomplish the same end result since the current implementation only prompts for missing input when the option was empty and the value returned from $this->laravel->getNamespace() was also empty; this proposed change would always prompt for the missing input when the option was empty regardless of the value returned by getNamespace().

LukeTowers commented:
Tests are failing because of a byproduct of the ochestra testbench setup, the scaffolded files are being generated inside of vendor/orchestra/testbench-core/laravel/$absolutePathToOrchestraTestBenchCoreLaravel; investigating further

LukeTowers committed:
Fix failing generator tests by registering App namespace in Orchestra Testbench

     Problem Context:
     -----------------
     The generator tests were failing because Orchestra Testbench creates a fake
     Laravel application in vendor/orchestra/testbench-core/laravel/ with its own
     composer.json that defines "App\\": "app/". However, the tests load Composer's
     autoloader from the framework's vendor/composer/autoload, which has no
     knowledge of the testbench app's namespace mappings.

     The new resolvePathForClass() method (introduced in the composer-based path
     resolution PR) relies on Composer's ClassLoader having registered namespace
     mappings to determine where to place generated files. When it can't resolve
     a namespace, it falls back to:

       $this->laravel['path.base'] . '/' . str_replace('\\', '/', $name) . '.php'

     This creates paths like:
       /vendor/.../laravel/App/View/Components/Foo.php
                           ^^^
     Instead of:
       /vendor/.../laravel/app/View/Components/Foo.php
                           ^^^

     The Problem with Test Generation:
     ----------------------------------
     When generators use the CreatesMatchingTest trait with the --test flag, the
     handleTestCreation() method attempts to extract the relative path using:

       (new Stringable($path))->after($this->laravel['path'])

     Where $this->laravel['path'] is "/vendor/.../laravel/app" (lowercase).

     This fails to find "/app" within "/App/View/Components/Foo.php" (uppercase),
     so the entire absolute path remains in the test name, creating deeply nested
     incorrect paths like:

       /vendor/.../laravel/tests/Feature/vendor/.../laravel/App/View/Components/FooTest.php

     The Solution:
     -------------
     Register the testbench app's App\\ namespace with Composer's ClassLoader in
     the test setUp() method. This allows resolvePathForClass() to succeed instead
     of falling back, generating correct paths that match the filesystem case.

     Alternative Approaches to Consider:
     -----------------------------------
     1. Adjust the fallback logic in getPath() to strip the root namespace and use
        $this->laravel['path'] (preserving pre-PR behavior), but this could break
        Winter CMS use cases where custom namespaces via --in don't match App\\.

     2. Make handleTestCreation() more robust by using case-insensitive matching
        or by extracting paths differently, though this doesn't address the root
        cause of the fallback being used unnecessarily.

     3. Have Orchestra Testbench itself register the App\\ namespace, though this
        would require changes in testbench-core and wouldn't help users on older
        versions.

     The current approach fixes the root cause in the test environment without
     affecting production behavior or breaking Winter CMS custom namespace support.

LukeTowers committed:
Merge branch '12.x' into patch-1

innocenzi commented:
@taylorotwell I know you personally prefer the classic structure, but for people working in other ways, that would be a nice quality of life to avoid fighting the framework! Could you consider it?

LukeTowers commented:
@taylorotwell I'm not sure how the infinite loop could be solved without also asking the user for more information on what namespace the class should be generated in. And if you're going to allow that anyways, I'm not sure why we wouldn't just allow it to be passed as an option from the start.

If you have ideas for how the code can better handle the case of an empty root namespace being set that would be appreciated. Do you have any concerns about how the namespace selection logic works? It will only ever prompt the user to provide one when there isn't a root namespace to use and won't change how most people interact with laravel/laravel applications at all.

crynobone committed:
Refactor setUp method to use default_skeleton_path

LukeTowers commented on tests/Integration/Generators/TestCase.php:
```diff
@@ -2,9 +2,21 @@
+        foreach (ClassLoader::getRegisteredLoaders() as $loader) {
+            $loader->addPsr4('App\\', [default_skeleton_path('app')]);
+        }
+        
```
```suggestion

```

LukeTowers committed:
Apply suggestion from @LukeTowers
